File: sdk/csharp/src/Providers/Tempmail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// tempmail.ing 渠道（api.tempmail.ing/api）。
/// 创建：POST /generate（duration 分钟）；收信：GET /emails/{email}，扁平化字段后归一。
/// </summary>
public static class Tempmail
{
    private const string Base = "https://api.tempmail.ing/api";

    private static Dictionary<string, string> Headers() => new()
    {
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36",
        ["Referer"] = "https://tempmail.ing/",
    };

    /// <summary>创建临时邮箱，支持自定义有效期（分钟）</summary>
    public static EmailInfo Generate(int duration)
    {
        var payload = Json.Serialize(new Dictionary<string, object?> { ["duration"] = duration });
        var resp = Http.Post($"{Base}/generate", payload, "application/json", Headers());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || (data["success"] as JsonValue)?.GetValue<bool>() != true)
            throw new Exception("Failed to generate email");
        var em = data["email"] as JsonObject ?? throw new Exception("Failed to generate email");
        return new EmailInfo("tempmail", Json.Str(em, "address"), createdAt: em["createdAt"]?.ToString());
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email)
    {
        var resp = Http.Get($"{Base}/emails/{WebUtility.UrlEncode(email)}", Headers());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject;
        if (data is null || (data["success"] as JsonValue)?.GetValue<bool>() != true)
            throw new Exception("Failed to get emails");
        var rows = data["emails"] as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var raw in rows)
        {
            if (raw is not JsonObject ro) continue;
            var from = Json.Str(ro, "from_address");
            if (string.IsNullOrEmpty(from)) from = Json.Str(ro, "from");
            var html = Json.Str(ro, "content");
            if (string.IsNullOrEmpty(html)) html = Json.Str(ro, "html");
            var date = Json.Str(ro, "received_at");
            if (string.IsNullOrEmpty(date)) date = Json.Str(ro, "date");
            var flat = new Dictionary<string, object?>
            {
                ["id"] = Json.Str(ro, "id"),
                ["from"] = from,
                ["to"] = email,
                ["subject"] = Json.Str(ro, "subject"),
                ["text"] = Json.Str(ro, "text"),
                ["html"] = html,
                ["date"] = date,
            };
            result.Add(Normalize.NormalizeEmail(flat, email));
        }
        return result;
    }
}

File: sdk/php/src/Providers/Tempmail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * tempmail.ing 渠道实现
 *
 * API: https://api.tempmail.ing/api
 * POST /generate 创建（支持有效期分钟数）；GET /emails/{email} 收信。
 */
final class Tempmail
{
    private const CHANNEL = 'tempmail';
    private const BASE_URL = 'https://api.tempmail.ing/api';

    /** @var array<string,string> */
    private const HEADERS = [
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36',
        'Content-Type' => 'application/json',
        'Referer' => 'https://tempmail.ing/',
        'DNT' => '1',
    ];

    /**
     * @param array<mixed> $raw
     * @return array<mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        $isRead = ($raw['is_read'] ?? null) === 1 || ($raw['is_read'] ?? null) === true;
        return [
            'id' => $raw['id'] ?? '',
            'from' => $raw['from_address'] ?? $raw['from'] ?? '',
            'to' => $recipient,
            'subject' => $raw['subject'] ?? '',
            'text' => $raw['text'] ?? '',
            'html' => $raw['content'] ?? $raw['html'] ?? '',
            'date' => $raw['received_at'] ?? $raw['date'] ?? '',
            'is_read' => $isRead,
            'attachments' => $raw['attachments'] ?? [],
        ];
    }

    public static function generate(int $duration = 30): EmailInfo
    {
        $resp = HttpClient::post(self::BASE_URL . '/generate', self::HEADERS, json: ['duration' => $duration]);
        $data = HttpClient::json($resp);
        if (empty($data['success']) || !is_array($data['email'] ?? null)) {
            throw new \RuntimeException('tempmail: 创建邮箱失败');
        }
        $email = $data['email'];
        $expiresAt = $email['expiresAt'] ?? null;
        return new EmailInfo(
            self::CHANNEL,
            (string) $email['address'],
            expiresAt: is_numeric($expiresAt) ? (int) $expiresAt : null,
            createdAt: $email['createdAt'] ?? null,
        );
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token = null): array
    {
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('tempmail: 邮箱地址为空');
        }
        $resp = HttpClient::get(self::BASE_URL . '/emails/' . rawurlencode($addr), self::HEADERS);
        $data = HttpClient::json($resp);
        if (empty($data['success'])) {
            throw new \RuntimeException('tempmail: 获取邮件失败');
        }
        $rows = $data['emails'] ?? [];
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $raw) {
            if (is_array($raw)) {
                $out[] = Normalize::email(self::flatten($raw, $addr), $addr);
            }
        }
        return $out;
    }
}
